src/service/resources-impl.js

```diff
@@ -1307,6 +1283,78 @@ export class Resources {
       }
     }
   }
+
+  /**
+   * Calls iterator on each sub-resource
+   * @param {function(!Resource, number)} iterator
+   */
+  setupVisibilityStateMachine_(vsm) {
+    const prerender = VisibilityState.PRERENDER;
+    const visible = VisibilityState.VISIBLE;
+    const hidden = VisibilityState.HIDDEN;
+    const paused = VisibilityState.PAUSED;
+    const inactive = VisibilityState.INACTIVE;
+
+    const doPass = () => {
+      // If viewport size is 0, the manager will wait for the resize event.
+      const viewportSize = this.viewport_.getSize();
+      if (viewportSize.height > 0 && viewportSize.width > 0) {
+        if (this.hasMutateWork_()) {
+          this.mutateWork_();
+        }
+        this.discoverWork_();
+        let delay = this.work_();
+        if (this.hasMutateWork_()) {
+          // Overflow mutate work.
+          delay = Math.min(delay, MUTATE_DEFER_DELAY_);
+        }
+        if (this.visible_) {
+          dev.fine(TAG_, 'next pass:', delay);
+          this.schedulePass(delay);
+          this.updateScrollHeight_();
+        } else {
+          dev.fine(TAG_, 'document is not visible: no scheduling');
+        }
+      }
+    };
+    const noop = () => {};
+    const pause = () => {
+      this.resources_.forEach(r => r.pause());
+    };
+    const unload = () => {
+      this.resources_.forEach(r => r.unload());
+    };
+    const resume = () => {
+      this.resources_.forEach(r => r.resume());
+      doPass();
+    };
+
+    vsm.addTransition(prerender, prerender, doPass);
+    vsm.addTransition(prerender, visible, doPass);
+    vsm.addTransition(prerender, hidden, doPass);
+    vsm.addTransition(prerender, inactive, doPass);
+    vsm.addTransition(prerender, paused, doPass);
+
+    vsm.addTransition(visible, visible, doPass);
+    vsm.addTransition(visible, hidden, doPass);
+    vsm.addTransition(visible, inactive, unload);
+    vsm.addTransition(visible, paused, pause);
+
+    vsm.addTransition(hidden, visible, doPass);
+    vsm.addTransition(hidden, hidden, doPass);
+    vsm.addTransition(hidden, inactive, unload);
+    vsm.addTransition(hidden, paused, pause);
+
+    vsm.addTransition(inactive, visible, doPass);
+    vsm.addTransition(inactive, hidden, doPass);
+    vsm.addTransition(inactive, inactive, noop);
+    vsm.addTransition(inactive, paused, doPass);
+
+    vsm.addTransition(paused, visible, resume);
+    vsm.addTransition(paused, hidden, doPass);
+    vsm.addTransition(paused, inactive, unload);
+    vsm.addTransition(paused, paused, doPass);
```

Comment:
The good thing about `unlayoutOnPause()` is that it will by us some time to resolve pausing on a case-by-case basis, instead of trying to capture all nuances in this PR. It could be a transition option.